scripts/fulfillable-roll.mjs

```diff
@@ -58,22 +58,56 @@ export async function _dieEvaluate(options) {
 
 /**
  * Identify terms in a Roll instance which are able to be externally fulfilled.
+ *
+ * Ignores terms which are already fulfilled.
+ *
  * @param {RollTerm[]} terms      Terms of the Roll instance
  * @param {object} config         The unfulfilled-rolls.diceSettings configuration
  * @returns {FulfillableTerm[]}   An array of identified terms
  * @private
  */
 function _identifyFulfillableTerms(terms, config) {
   const toFulfill = [];
-  for ( const [i, term] of terms.entries() ) {
+  const diceTerms = terms.flatMap(t => _diceFromRollTerm(t, !!config.searchNestedDice));
```

Comment:
The idea here is right, we need to iterate over dice terms and recursively handle terms which contain inner dice. 

This implementation fully iterates first, however, to create `const diceTerms` rather than recursively evaluating terms as we go. This results in us iterating over all terms twice, instead of just once.

A better way to do this will be something like this:

```js
function function _identifyFulfillableTerms(terms, config, toFulfill=[]) {
  for ( const [i, term] of diceTerms.entries() ) {
    if ( !(term instanceof Die) ) continue;
    if ( this._fulfilled?.length > 0 ) continue; // already fulfilled
    const method = config[`d${term.faces}`];
    if ( !method || (method === "fvtt") ) continue;
    toFulfill.push({term, method, index: i});

    // Recursively identify inner terms
    if ( "dice" in term ) {
      const innerTerms = Array.from(term.dice);
      _identifyFulfillableTerms(innerTerms, config, toFulfill);
    }
  }
  return toFulfill;
}
```

Reply:
happy to change that.

Reply:
I'd like to avoid modifying the interface of this function though, I think it'll probably be better to make the extraction inside, avoiding exposing `toFulfill` as an argument. I'll change the PR.

Reply:
It's a private function, change whatever you need to change to have the cleanest implementation.